src/cascadia/TerminalSettingsEditor/NewTabMenuViewModel.cpp

```diff
@@ -0,0 +1,768 @@
+                CurrentView().InsertAt(idx + 1, vm);
+            }
+        }
+    }
+
+    void NewTabMenuViewModel::RequestDeleteEntry(const Editor::NewTabMenuEntryViewModel& vm)
+    {
+        uint32_t idx;
+        if (CurrentView().IndexOf(vm, idx))
+        {
+            CurrentView().RemoveAt(idx);
+
+            if (vm.try_as<Editor::FolderEntryViewModel>())
+            {
+                _folderTreeCache = nullptr;
+            }
+        }
+    }
+
+    void NewTabMenuViewModel::RequestMoveEntriesToFolder(const Windows::Foundation::Collections::IVector<Editor::NewTabMenuEntryViewModel>& entries, const Editor::FolderEntryViewModel& destinationFolder)
+    {
+        for (auto&& e : entries)
+        {
+            // Remove entry from the current layer,
+            // and add it to the destination folder
+            RequestDeleteEntry(e);
+            destinationFolder.Entries().Append(e);
+        }
+    }
+
+    void NewTabMenuViewModel::RequestAddSelectedProfileEntry()
+    {
+        if (_SelectedProfile)
+        {
+            Model::ProfileEntry profileEntry;
+            profileEntry.Profile(_SelectedProfile);
+
+            CurrentView().Append(make<ProfileEntryViewModel>(profileEntry));
+        }
+        _PrintAll();
+    }
+
+    void NewTabMenuViewModel::RequestAddSeparatorEntry()
+    {
+        Model::SeparatorEntry separatorEntry;
+
+        CurrentView().Append(make<SeparatorEntryViewModel>(separatorEntry));
+
+        _PrintAll();
+    }
+
+    void NewTabMenuViewModel::RequestAddFolderEntry()
+    {
+        Model::FolderEntry folderEntry;
+        folderEntry.Name(_AddFolderName);
+
+        CurrentView().Append(make<FolderEntryViewModel>(folderEntry, _Settings));
+
+        // Reset state after adding the entry
+        AddFolderName({});
+        _folderTreeCache = nullptr;
+
+        _PrintAll();
+    }
+
+    void NewTabMenuViewModel::RequestAddProfileMatcherEntry()
+    {
+        Model::MatchProfilesEntry matchProfilesEntry;
+        matchProfilesEntry.Name(_ProfileMatcherName);
+        matchProfilesEntry.Source(_ProfileMatcherSource);
+        matchProfilesEntry.Commandline(_ProfileMatcherCommandline);
+
+        CurrentView().Append(make<MatchProfilesEntryViewModel>(matchProfilesEntry));
+
+        // Clear the fields after adding the entry
+        ProfileMatcherName({});
+        ProfileMatcherSource({});
+        ProfileMatcherCommandline({});
+
+        _PrintAll();
+    }
+
+    void NewTabMenuViewModel::RequestAddRemainingProfilesEntry()
+    {
+        Model::RemainingProfilesEntry remainingProfilesEntry;
+        CurrentView().Append(make<RemainingProfilesEntryViewModel>(remainingProfilesEntry));
+
+        _NotifyChanges(L"IsRemainingProfilesEntryMissing");
+
+        _PrintAll();
+    }
+
+    void NewTabMenuViewModel::GenerateFolderTree()
+    {
+        if (!_folderTreeCache)
+        {
+            _folderTreeCache = single_threaded_observable_vector<Editor::FolderTreeViewEntry>();

```

Comment:
Not sure what you mean here. Switching over to an std::vector would complicate things imo.`_folderTreeCache` only holds the root, and we create the tree by looking at FolderEntry's children (which we wouldn't be able to look at anyways since that's forced to be an IVector).

Reply:
You can do it like this:
```cpp
void NewTabMenuViewModel::GenerateFolderTree()
{
    if (!_folderTreeCache)
    {
        // Add the root folder
        auto root = winrt::make<FolderTreeViewEntry>(nullptr);

        for (const auto&& entry : _rootEntries)
        {
            if (entry.Type() == NewTabMenuEntryType::Folder)
            {
                root.Children().Append(winrt::make<FolderTreeViewEntry>(entry.as<Editor::FolderEntryViewModel>()));
            }
        }

        std::vector<Editor::FolderTreeViewEntry> folderTreeCache;
        folderTreeCache.emplace_back(std::move(root));
        _folderTreeCache = single_threaded_observable_vector(std::move(folderTreeCache));

        _NotifyChanges(L"FolderTree");
    }
}
```

But it doesn't matter much.